File: sda/src/Database/Transaction.cpp
#include "SDA/Database/Transaction.h"
#include "SDA/Database/Database.h"

using namespace sda;

Transaction::Transaction(Database* database)
    : m_database(database)
{}

void Transaction::markAsNew(utils::ISerializable* obj) {
    if (m_objects.find(obj) != m_objects.end())
        throw std::runtime_error("Object already marked as new");
    m_objects[obj] = New;
}

void Transaction::markAsModified(utils::ISerializable* obj) {
    auto it = m_objects.find(obj);
    if (it != m_objects.end()) {
        if (it->second == Removed)
            throw std::runtime_error("Object already marked as removed");
        return;
    }
    m_objects[obj] = Modified;
}

void Transaction::markAsRemoved(utils::ISerializable* obj) {
    auto it = m_objects.find(obj);
    if (it != m_objects.end()) {
        if (it->second == New) {
            // the object is new, so we can just remove it from the transaction and don't save it
            m_objects.erase(it);
            return;
        }
    }
    m_objects[obj] = Removed;
}

void Transaction::commit() {
    for(auto& [obj, changeType] : m_objects) {
        boost::json::object data;
        obj->serialize(data);
        if (data["temporary"].get_bool())
            continue;
        auto collectionName = std::string(data["collection"].get_string().c_str());
        auto collection = m_database->getCollection(collectionName);
        if (changeType == New || changeType == Modified) {
            collection->write(data);
        } else if (changeType == Removed) {
            collection->remove(data);
        }
    }

    m_objects.clear();
}

File: javascript/packages/sda-electron/src/database/transaction.ts
import { Database, IDatabaseObject } from './database';

enum ChangeType {
  New,
  Modified,
  Removed,
}

type Change = {
  type: ChangeType;
  obj: IDatabaseObject;
};

export class Transaction {
  private readonly db: Database;
  private changes: { [id: string]: Change } = {};

  constructor(db: Database) {
    this.db = db;
  }

  markAsNew(obj: IDatabaseObject) {
    if (this.changes[obj.id]) {
      throw new Error(`Object ${obj.id} is already marked as new`);
    }
    this.changes[obj.id] = { type: ChangeType.New, obj };
  }

  markAsModified(obj: IDatabaseObject) {
    const change = this.changes[obj.id];
    if (change) {
      if (change.type === ChangeType.Removed) {
        throw new Error(`Object ${obj.id} is marked as removed`);
      }
    } else {
      this.changes[obj.id] = { type: ChangeType.Modified, obj };
    }
  }

  markAsRemoved(obj: IDatabaseObject) {
    const change = this.changes[obj.id];
    if (change) {
      if (change.type === ChangeType.Removed) {
        throw new Error(`Object ${obj.id} is already marked as removed`);
      } else if (change.type === ChangeType.New) {
        delete this.changes[obj.id];
        return;
      }
    }
    this.changes[obj.id] = { type: ChangeType.Removed, obj };
  }

  async commit() {
    for (const [_, change] of Object.entries(this.changes)) {
      const data = change.obj.serialize();
      if (change.type === ChangeType.New || change.type === ChangeType.Modified) {
        await this.db.upsert(data);
      } else if (change.type === ChangeType.Removed) {
        await this.db.delete(data);
      }
    }
    this.changes = {};
  }

  isEmpty() {
    return Object.keys(this.changes).length === 0;
  }
}
